### ISSUE TITLE
nvidia-smi utilization and memory stats native in pytorch

### ISSUE THREAD

**Issue by msaroufim** [pytorch/pytorch#65870]
## 🚀 Feature

GPU utilization and memory usage stats natively in pytorch

## Motivation

Almost every deep learning engineer I know runs `watch -n0.1 nvidia-smi` in the background to answer 2 questions
* Was a model loaded to my GPU? By looking at the on device memory
* What's the GPU utilization? To answer whether they should increase the batch size

Here's a stack overflow thread asking "How do I know if my GPU being used" where the top voted answer has over 450 upvotes!! https://stackoverflow.com/questions/48152674/how-to-check-if-pytorch-is-using-the-gpu

Answering whether a GPU is being used or not throws off beginners and advanced users alike who will run things like

```
import torch
torch.cuda.is_available()
```

And then `device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")`

And need to double check that all tensors are on the GPU by writing something like `tensor.to(device)` since Pytorch doesn't send tensors to GPUs by default if its available

And ultimately you either know if your GPU is being used or not by either checking if a model feels too slowly or going to `nvidia-smi`

## Pitch

```
`torch.cuda.utilization() -> 99%
`torch.cuda.memory_usage() -> 30%
```

`nvidia-smi` is great but people need this data at runtime in their code, if a model wasn't loaded to GPU stop. Or GPU utlization is too low, restart job with a larger batch size

We could just do a grep search on top `os.system('nvidia-smi')` and capture that in functions or maybe there's some fancier way to do this that doesns't involve grepping

Maybe this needs more thought in distributed settings but for now we can limit scope to single node

## Alternatives

This gap is already met by a bunch of tools like
* https://github.com/anderskm/gputil
* https://github.com/rossumai/nvgpu

But those tools provide comprehensive information. We just need to help deep learning practitioners  debug their issues namely: Are you using your GPU? And are you using your GPU to its full capacity?

@dzhulgakov @gdankel @ngimel 

## Additional context

![image](https://user-images.githubusercontent.com/3282513/135367198-fc750f17-d6e0-48f8-97fc-1122f4dedafd.png)

cc @ngimel

**Comment by alexeygolyshev** [comment]
Plus `torch.cuda.temperature()`

**Comment by ngimel** [comment]
Can be easily done with `pynvml.nvmlDeviceGetUtilizationRates(handle)`. 
However, keep in mind that in cases of multiple processes running on the same gpu this may not be the metric you want (e.g. some other process is running on the gpu, and your process is not using gpu at all).

**Comment by pankajtakawale** [comment]
looking forward for this change

**Comment by ngimel** [comment]
Added in #69104
